#Name : sanket B.
#PRN : 19060641032

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt 
%matplotlib inline 

In [2]:
df = pd.read_csv('Attrition.csv')
df.head()

Attrition  Age     BusinessTravel  DailyRate              Department  \
0          1   41      Travel_Rarely        802                   Sales   
1          0   49  Travel_Frequently        279  Research & Development   
2          1   36      Travel_Rarely       1373  Research & Development   
3          0   33  Travel_Frequently       1392  Research & Development   
4          0   27      Travel_Rarely        591  Research & Development   

   DistanceFromHome EducationField  Gender  HourlyRate                JobRole  \
0                 1  Life Sciences  Female          94        Sales Executive   
1                 8  Life Sciences    Male          61     Research Scientist   
2                 2          Other    Male          66  Laboratory Technician   
3                 3  Life Sciences  Female          56     Research Scientist   
4                 2        Medical    Male          40  Laboratory Technician   

   JobSatisfaction  MonthlyIncome  NumCompaniesWorked OverTime  \
0                4           5993                   8      Yes   
1                2           5130                   1       No   
2                3           2090                   6      Yes   
3                3           2909                   1      Yes   
4                2           3468                   9       No   

   PercentSalaryHike  PerformanceRating  WorkLifeBalance  YearsAtCompany  \
0                 11                  3                1               6   
1                 23                  4                3              10   
2                 15                  3                3               0   
3                 11                  3                3               8   
4                 12                  3                3               2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [3]:
df.shape

(1263, 20)

In [4]:
df.shape
df.isna().value_counts()

Attrition  Age    BusinessTravel  DailyRate  Department  DistanceFromHome  EducationField  Gender  HourlyRate  JobRole  JobSatisfaction  MonthlyIncome  NumCompaniesWorked  OverTime  PercentSalaryHike  PerformanceRating  WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager
False      False  False           False      False       False             False           False   False       False    False            False          False               False     False              False              False            False           False                    False                   1263
dtype: int64

In [5]:
df1 = df.drop(['BusinessTravel','Department','EducationField','Gender','JobRole','OverTime','Gender'],axis=1)
df1.head()

Attrition  Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0          1   41        802                 1          94                4   
1          0   49        279                 8          61                2   
2          1   36       1373                 2          66                3   
3          0   33       1392                 3          56                3   
4          0   27        591                 2          40                2   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0           5993                   8                 11                  3   
1           5130                   1                 23                  4   
2           2090                   6                 15                  3   
3           2909                   1                 11                  3   
4           3468                   9                 12                  3   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0                1               6                        0   
1                3              10                        1   
2                3               0                        0   
3                3               8                        3   
4                3               2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2

In [6]:
#df.head()
x = df1.drop('Attrition',axis=1)
y = df1['Attrition']

In [7]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.30,random_state=101)

In [8]:
#from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
df1 = RandomForestClassifier()
df1.fit(x_train, y_train)

RandomForestClassifier()

In [9]:
df1_pred = df1.predict(x_test)

### F1 Score

In [10]:
from sklearn.metrics import f1_score
df1_data=(f1_score(y_test, df1_pred))
df1_data

0.08823529411764705

## Drop 10% data Randomly from numeric columns.

In [11]:
df2 = df.drop(['BusinessTravel','Department','EducationField','Gender','JobRole','OverTime','Gender','Attrition'],axis=1)
df2.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0   41        802                 1          94                4   
1   49        279                 8          61                2   
2   36       1373                 2          66                3   
3   33       1392                 3          56                3   
4   27        591                 2          40                2   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0           5993                   8                 11                  3   
1           5130                   1                 23                  4   
2           2090                   6                 15                  3   
3           2909                   1                 11                  3   
4           3468                   9                 12                  3   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0                1               6                        0   
1                3              10                        1   
2                3               0                        0   
3                3               8                        3   
4                3               2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2

In [12]:
for col in df2.columns:
    df2.loc[df2.sample(frac=0.1).index,col]=np.nan

In [13]:
df2.isnull().sum()

Age                        126
DailyRate                  126
DistanceFromHome           126
HourlyRate                 126
JobSatisfaction            126
MonthlyIncome              126
NumCompaniesWorked         126
PercentSalaryHike          126
PerformanceRating          126
WorkLifeBalance            126
YearsAtCompany             126
YearsSinceLastPromotion    126
YearsWithCurrManager       126
dtype: int64

In [14]:
df2_median = df2.copy()
df2_mean = df2.copy()
df2_min = df2.copy()
df2_max = df2.copy()
df2_mode=df2.copy()

In [15]:
df2.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                      NaN   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

## Mean Imputation

In [16]:
df2_mean.columns
for i in list(df2_mean.columns):
    df2_mean[i].fillna(int(df2_mean[i].mean()),inplace=True)
df2_mean.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                      1.0   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

## Median Imputation

In [17]:
df2_median.columns
for i in list(df2_median.columns):
    df2_median[i].fillna(int(df2_median[i].median()),inplace=True)
df2_median.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                      1.0   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

## Mode Imputation

In [18]:
df2_mode.columns
for i in list(df2_mode.columns):
    df2_mode[i].fillna(int(df2_mode[i].mode().mean()),inplace=True)
df2_mode.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                      0.0   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

In [19]:
df2_min.columns
for i in list(df2_min.columns):
    df2_min[i].fillna(int(df2_min[i].min()),inplace=True)
df2_min.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                      0.0   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

## Max Imputation

In [20]:
df2_max.columns
for i in list(df2_max.columns):
    df2_max[i].fillna(int(df2_max[i].max()),inplace=True)
df2_max.head()

Age  DailyRate  DistanceFromHome  HourlyRate  JobSatisfaction  \
0  41.0      802.0               1.0        94.0              4.0   
1  49.0      279.0               8.0        61.0              2.0   
2  36.0     1373.0               2.0        66.0              3.0   
3  33.0     1392.0               3.0        56.0              3.0   
4  27.0      591.0               2.0        40.0              2.0   

   MonthlyIncome  NumCompaniesWorked  PercentSalaryHike  PerformanceRating  \
0         5993.0                 8.0               11.0                3.0   
1         5130.0                 1.0               23.0                4.0   
2         2090.0                 6.0               15.0                3.0   
3         2909.0                 1.0               11.0                3.0   
4         3468.0                 9.0               12.0                3.0   

   WorkLifeBalance  YearsAtCompany  YearsSinceLastPromotion  \
0              1.0             6.0                      0.0   
1              3.0            10.0                     15.0   
2              3.0             0.0                      0.0   
3              3.0             8.0                      3.0   
4              3.0             2.0                      2.0   

   YearsWithCurrManager  
0                   5.0  
1                   7.0  
2                   0.0  
3                   0.0  
4                   2.0

## Model Fitting on Mean Imputed data

In [21]:
x_mean = df2_mean[:]
y_mean = df['Attrition']

In [22]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_mean_train,x_mean_test,y_mean_train,y_mean_test = train_test_split(x_mean,y_mean,test_size=0.30,random_state=101)

In [23]:
from sklearn.ensemble import RandomForestClassifier
df2 =  RandomForestClassifier() 
df2.fit(x_mean_train, y_mean_train)

RandomForestClassifier()

In [24]:
df2_pred = df2.predict(x_mean_test)

#### F1 Score

In [25]:
from sklearn.metrics import f1_score
df2_data=(f1_score(y_mean_test, df2_pred))
df2_data

0.11267605633802817

## Model Fitting on Median Imputed data

In [26]:
x_median = df2_median[:]
y_median = df['Attrition']

In [27]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_median_train,x_median_test,y_median_train,y_median_test = train_test_split(x_median,y_median,test_size=0.30,random_state=101)

In [28]:
from sklearn.ensemble import RandomForestClassifier
df3 =  RandomForestClassifier() 
df3.fit(x_median_train, y_median_train)

RandomForestClassifier()

In [29]:
df3_pred = df2.predict(x_median_test)

### F1 Score

In [30]:
from sklearn.metrics import f1_score
df3_data=(f1_score(y_median_test, df3_pred))
df3_data

0.08823529411764705

## Model Fitting on Mode Imputed data

In [31]:
x_mode = df2_mode[:]
y_mode = df['Attrition']

In [32]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_mode_train,x_mode_test,y_mode_train,y_mode_test = train_test_split(x_mode,y_mode,test_size=0.30,random_state=101)

In [33]:
from sklearn.ensemble import RandomForestClassifier
df4 =  RandomForestClassifier() 
df4.fit(x_mode_train, y_mode_train)

RandomForestClassifier()

In [34]:
df4_pred = df2.predict(x_mode_test)

### F1 Score

In [35]:
from sklearn.metrics import f1_score
df4_data=(f1_score(y_mode_test, df4_pred))
df4_data

0.05970149253731343

## Model Fitting on Min Imputed data

In [36]:
x_min = df2_min[:]
y_min = df['Attrition']

In [37]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_min_train,x_min_test,y_min_train,y_min_test = train_test_split(x_min,y_min,test_size=0.30,random_state=101)

In [38]:
from sklearn.ensemble import RandomForestClassifier
df5 =  RandomForestClassifier() 
df5.fit(x_min_train, y_min_train)

RandomForestClassifier()

In [39]:
df5_pred = df2.predict(x_min_test)

###  F1 Score



In [40]:
from sklearn.metrics import f1_score
df5_data=(f1_score(y_min_test, df5_pred))
df5_data

0.12121212121212122

## Model Fitting on Max Imputed data

In [41]:
x_max = df2_max[:]
y_max = df['Attrition']

In [42]:
# fit a trasformation
from sklearn.model_selection import train_test_split
x_max_train,x_max_test,y_max_train,y_max_test = train_test_split(x_max,y_max,test_size=0.30,random_state=101)

In [43]:
from sklearn.ensemble import RandomForestClassifier
df6 =  RandomForestClassifier() 
df6.fit(x_max_train, y_max_train)

RandomForestClassifier()

In [44]:
df6_pred = df2.predict(x_max_test)

### F1 Score

In [45]:
from sklearn.metrics import f1_score
df6_data=(f1_score(y_min_test, df6_pred))
df6_data

0.13513513513513511

### After removing 10% of the data randomly and imputed by different methods (mean,median,mode,min,max) the decision tree model was reassembled to the imputed data, then f1 scores as follows by different methods.
* mean imputation F1 Score is 0.11267605
* median imputation F1 Score is 0.08823529 
* mode imputation F1 Score is 0.05970149
* min imputation F1 Score is 0.1212121
* max imputation F1 Score is 0.1351351
### As we can see, the  F1 Score of maximum imputation is higher than other imputations.